File: problem6/dcp6.rb
class Node
    attr_reader :val
    attr_accessor :npx

    def initialize data
        @val = data
        @npx = 0
    end
end

class XORLinkedList
    attr_reader :root, :tail, :nodes

    def initialize data
        @root = Node.new(data)
        @nodes = 1
    end

    def add element
        new_node = Node.new(element)

        if @root.npx == 0
            @root.npx = get_pointer(new_node)
            new_node.npx = get_pointer(@root)
        elsif
            curr_node = deref_pointer(@root.npx)
            prev_addr = get_pointer(@root)
            
            while curr_node.npx != prev_addr
                next_node = deref_pointer(curr_node.npx ^ prev_addr)
                prev_addr = get_pointer(curr_node)
                curr_node = next_node
            end

            new_node.npx = get_pointer(curr_node)
            curr_node.npx = prev_addr ^ get_pointer(new_node)
        end

        @nodes += 1
    end

    def get(index)
        if index == 0
            return @root.val
        end

        curr = deref_pointer(@root.npx)
        prev_addr = get_pointer(@root)

        for i in (1 ... index) do
            next_node = deref_pointer(prev_addr ^ curr.npx)   
            prev_addr = get_pointer(curr)
            curr = next_node
        end

        curr.val
    end
end

def get_pointer(node)
    node.object_id
end

def deref_pointer(node_addr)
    ObjectSpace._id2ref(node_addr)
end

list = XORLinkedList.new("A")
list.add("B")
list.add("C")
list.add("D")
list.add("E")

puts list.get(0)
puts list.get(1)
puts list.get(2)
puts list.get(3)
puts list.get(4)

File: problem6/dcp6.c
#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>

typedef struct node
{
    uintptr_t both;
    char val;
} Node;

typedef struct XORLinkedList
{
    Node *root;
} xorlist;

void add(xorlist *list, const char val)
{
    Node *new_node = malloc(sizeof(*new_node));
    new_node->both = 0;
    new_node->val = val;

    /* 0 nodes */
    if (list->root == NULL)
    {
        list->root = new_node;
    }   
    /* 1 node */
    else if (list->root->both == 0)
    {
        list->root->both = (uintptr_t)new_node;
        new_node->both = (uintptr_t)list->root;
    }
    /* > 1 nodes */
    else
    {
        Node *curr_node = (Node *)list->root->both;
        uintptr_t prev_addr = (uintptr_t)list->root;

        while (curr_node->both != prev_addr)
        {
            Node *next_node = (Node *)(prev_addr ^ curr_node->both);
            prev_addr = (uintptr_t)curr_node;
            curr_node = next_node;
        }

        curr_node->both = prev_addr ^ (uintptr_t)new_node;
        new_node->both = (uintptr_t)curr_node;
    }
}

char get(xorlist *list, const size_t pos)
{
    if (pos == 0)
    {
        return list->root->val;
    }

    Node *curr_node = (Node *)list->root->both;
    uintptr_t prev_addr = (uintptr_t)list->root;

    for(size_t i=1; i<pos; ++i)
    {
        Node *next_node = (Node *)(prev_addr ^ curr_node->both);
        prev_addr = (uintptr_t)curr_node;
        curr_node = next_node;
    }

    return curr_node->val;
}

int main(void)
{
    xorlist *list = malloc(sizeof(*list));
    list->root = NULL;
    add(list, 'A');
    add(list, 'B');
    add(list, 'C');
    add(list, 'D');
    add(list, 'E');

    printf("%c\n", get(list, 0));
    printf("%c\n", get(list, 1));
    printf("%c\n", get(list, 2));
    printf("%c\n", get(list, 3));
    printf("%c\n", get(list, 4));

    return 0;
}
